src/thread/thread.cpp

```diff
@@ -62,59 +62,58 @@ namespace fc {
   }
 
    thread*& current_thread() {
-      #ifdef _MSC_VER
+#ifdef _MSC_VER
          static __declspec(thread) thread* t = NULL;
-      #else
+#else
          static __thread thread* t = NULL;
-      #endif
+#endif
       return t;
    }
 
-   thread::thread( const std::string& name  ) {
+   thread::thread( const std::string& name, thread_idle_notifier* notifier ) {
       promise<void>::ptr p(new promise<void>("thread start"));
-      boost::thread* t = new boost::thread( [this,p,name]() {
+      boost::thread* t = new boost::thread( [this,p,name,notifier]() {
           try {
             set_thread_name(name.c_str()); // set thread's name for the debugger to display
-            this->my = new thread_d(*this);
+            this->my = new thread_d( *this, notifier );
             current_thread() = this;
             p->set_value();
             exec();
           } catch ( fc::exception& e ) {
-            wlog( "unhandled exception" );
-            p->set_exception( e.dynamic_copy_exception() );
+            if( !p->ready() )
+            {
+               wlog( "unhandled exception" );
+               p->set_exception( e.dynamic_copy_exception() );
+            }
+            else
+            { // possibly shutdown?
+               std::cerr << "unhandled exception in thread '" << name << "'\n";
+               std::cerr << e.to_detail_string( log_level::warn );
+            }
           } catch ( ... ) {
-            wlog( "unhandled exception" );
-            p->set_exception( std::make_shared<unhandled_exception>( FC_LOG_MESSAGE( warn, "unhandled exception: ${diagnostic}", ("diagnostic",boost::current_exception_diagnostic_information()) ) ) );
-            //assert( !"unhandled exception" );
-            //elog( "Caught unhandled exception %s", boost::current_exception_diagnostic_information().c_str() );
+            if( !p->ready() )
+            {
+               wlog( "unhandled exception" );
+               p->set_exception( std::make_shared<unhandled_exception>( FC_LOG_MESSAGE( warn, "unhandled exception: ${diagnostic}", ("diagnostic",boost::current_exception_diagnostic_information()) ) ) );
+            }
+            else
+            { // possibly shutdown?
+               std::cerr << "unhandled exception in thread '" << name << "'\n";
```

Comment:
To compile on Ubuntu 18.04, need to include iostream or remove stdout streams

Reply:
Added. Thanks!